### PR TITLE
[Perf] Metapath random walk with OpenMP

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Speeding up (?) metapath random walk sampler with OpenMP.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

File: src/graph/sampler/metapath.cc [hunk 1]
```diff
@@ -56,14 +65,27 @@ RandomWalkTracesPtr MetapathRandomWalk(
...
     }
 
-    trace_counts.push_back(curr_num_traces);
+    vertices_per_seed[seed_id] = curr_vertices;
+    trace_lengths_per_seed[seed_id] = curr_trace_lengths;
+    trace_counts[seed_id] = curr_num_traces;
+  }
+
+  for (uint64_t seed_id = 0; seed_id < num_seeds; ++seed_id) {
+    vertices.insert(
```
Comment (classicsong) [inline src/graph/sampler/metapath.cc]: Can we pre-alloc the vertices and trace_lengths?

Comment (BarclayII) [reply src/graph/sampler/metapath.cc]: No.  The lengths of traces are not always the same.

File: src/graph/sampler/metapath.cc [hunk 2]
```diff
@@ -39,11 +39,20 @@ RandomWalkTracesPtr MetapathRandomWalk(
...
+  std::vector< std::vector<size_t> > trace_lengths_per_seed(num_seeds);
+  // trace_counts[seed]
+  std::vector<size_t> trace_counts(num_seeds);
   std::vector<dgl_id_t> vertices;
-  std::vector<size_t> trace_lengths, trace_counts;
+  std::vector<size_t> trace_lengths;
 
   // TODO(quan): use omp to parallelize this loop
-  for (uint64_t seed_id = 0; seed_id < num_seeds; ++seed_id) {
+#pragma omp parallel for
```
Comment (jermainewang) [inline src/graph/sampler/metapath.cc]: Be aware of `hg->SuccVec`. It will materialize CSR if it does not exist and cause data race.

Comment (BarclayII) [reply src/graph/sampler/metapath.cc]: Is it doing so now?  I couldn't find any reference to lazy materialization of CSR for heterographs.

**Comment by BarclayII** [comment]
Temporarily closing this since the speed up is not working as expected.

**Pull request closed without merging**